---
title: STEAM
author: Kelsey Grinde
date: `r Sys.Date()`
output:
  md_document:
    variant: markdown_github
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r, echo = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "README-"
)
```

# Significance Threshold Estimation for Admixture Mapping

*STEAM* (Significance Threshold Estimation for Admixture Mapping) is an R package for estimating genome-wide significance thresholds for admixture mapping studies. 

**This package is under active development, so please stay tuned for future updates!**

# Citation

If you use *STEAM*, please cite the following article:

Grinde, K., Brown, L., Reiner, A., Thornton, T., & Browning, S. (TBD). Genome-wide significance thresholds for admixture mapping studies. *(This manuscript is currently under review. Citation information will be updated once published.)* 

# Installation

This package is free to use, under the terms of the MIT License. For more details, see LICENSE and [here](https://opensource.org/licenses/MIT).

To install *STEAM*, run the following commands in `R`:

```{r download, eval = FALSE}
##install.packages("devtools") # run this line if you have not yet installed the devtools package
library(devtools)
install_github("kegrinde/STEAM")
```

# Using *STEAM* to Estimate Significance Thresholds

In Grinde et al. (TBD), we propose two approaches for estimating genome-wide significance thresholds for admixture mapping studies:

- **Analytic Approximation:** applicable to admixed populations with 2 ancestral populations
- **Test Statistic Simulation:** applicable to admixed populations with 2 or more ancestral populations

To run either approach, we first need to:

1. Create a `map` file containing, at minimimum, the chromosome number and genetic position (in centimorgans) of each marker being tested.
    - *NOTE:* If you used a local ancestry inference program that performs calling within windows (e.g., RFMix), we recommend that you use just a single marker per window.
2. Estimate the admixture proportions for each individual and store them in a `n`-by-`K` matrix, where `n` is the number of admixed individuals and `K` is the number of ancestral populations.
    - There are various ways to calculate these proportions, one of which is to calculate the genome-wide average local ancestry for each individual.
3. Estimate `g`, the number of generations since admixture.
    - We recommend that you use *STEAM* for this step; for more details, see the "Estimating the Number of Generations..." section below.

## Example: 2 Ancestral Populations 

For an admixture mapping study in an admixed population with two ancestral populations (e.g., African Americans), we can use either approach (analytic approximation or test statistic simulation) to estimate the genome-wide significance threshold for our study. The analytic approximation approach is the faster of the two options.

Suppose we have markers spaced every 0.2 cM across 22 chromosomes. We store the information about genetic position and chromosome for each marker in a data frame called `example_map`:

```{r example_map}
head(example_map)
```

Suppose as well that the individuals in our sample have admixture proportions that are uniformly distributed from 0 to 1. We store these proportions in a data frame called `example_props`:

```{r example_props}
head(example_props)
```

We can use *STEAM* to estimate the number of generations since admixture (`g`) based on the observed pattern of correlation in local ancestry at pairs of markers across the genome. (Code will be posted soon). Suppose we estimate this value to be 6. 

We wish to estimate the *p*-value threshold which will control the family-wise error rate for this study at the 0.05 level. Since we have two ancestral populations, we can use either the analytic approximation or test statistic simulation approach.

### Analytic Approximation

To implement the analytic approximation approach, use the following command:

```{r example-analytic}
get_thresh_analytic(g = 6, map = example_map, type = "pval")
```

### Test Statistic Simualtion

To implement the test statistic simulation approach, we need to specify the number of replications for the simulation study (`nreps`). Computation time increases with the number of replications, so for the purposes of this example we choose a small number of reps. In practice, we recommend using a much larger number number of replications (the *STEAM* default is 10000). 

The `R` command for the test statistic simulation approach looks like this:

```{r example-simstat}
set.seed(1) # set seed for reproducibility
get_thresh_simstat(g = 6, map = example_map, props = example_props, nreps = 50)
```

Note that this approach provides both an estimate of the significance threshold and a 95\% bootstrap confidence interval for that threshold.

## Example: 3 Ancestral Populations 

For an admixture mapping study in an admixed population with three or more ancestral populations (e.g., Hispanics/Latinos), the analytic approximation is no longer applicable. However, we can still use the test statistic simulation approach to estimate the genome-wide significance threshold for our study. 

Suppose, as in the previous example, we have markers spaced every 0.2 cM across 22 chromosomes. As before, we store the information about genetic position and chromosome for each marker in a data frame called `example_map`.

Now, suppose that the individuals have genetic material contributed from three ancestral populations. We estimate admixture proportions and store them in a data frame called `example_props_K3`:

```{r example-props-K3}
head(example_props_K3)
```

As in the case of 2 ancestral populations, we can use *STEAM* to estimate the number of generations since admixture (`g`) based on the observed pattern of correlation in local ancestry at pairs of markers across the genome. (Code will be posted soon). Suppose we estimate this value to be 10. 

To estimate the *p*-value threshold which will control the family-wise error rate for this study at the 0.05 level, we run the following command:

```{r example-simstat-K3}
set.seed(1) # set seed for reproducibility
get_thresh_simstat(g = 10, map = example_map, props = example_props_K3, nreps = 50)
```

# Important Considerations

## Admixture Mapping Model Choice

Our multiple testing correction procedures assume that admixture mapping is being performed using a marginal regression approach, regressing the trait on local ancestry for each marker and each ancestral population one-by-one. Importantly, these regression models should include admixture proportions as covariates. For more details, see Grinde et al. (TBD).

## Estimating the Number of Generations since Admixture

Coming soon!
